File: descriptors/python_ctypes/test_library.f90
subroutine hello_world()
  write(*,*) "Hello, world!"
  return
end subroutine hello_world

subroutine create_vector(n, myvec)
  integer, intent(in), value :: n
  double precision, intent(out) :: myvec(n)
  double precision fac

  fac = 100.0d0/n
  
  do i = 1, n
     myvec(i) = fac*i
  end do
  
  return 
end subroutine create_vector

subroutine create_matrix(n1, n2, array)
  integer, intent(in), value :: n1
  integer, intent(in), value :: n2
  ! note transposed dimensions
  double precision, intent(out) :: array(n2, n1)
  double precision fac

  fac = 1.0d0/(n1*(n2+1))
  
  do i = 1, n1
     do j = 1, n2
        ! note transposed indices
        array(j,i) = fac*i*(j+1)
     end do
  end do
  
  return 
end subroutine create_matrix

File: ml-dft-sandia/descriptors/python_ctypes/test_library.cpp
#include <cstdio>
#include <cstdlib>
#include "test_library.h"

void hello_world() {
  printf(" Hello, world!\n");
}

double** create_matrix(int n1, int n2) {
  double **array;
  double *values;

  values = new double[n1*n2];
  array = new double*[n1];
  
  int n = 0;
  for (int i = 0; i < n1; i++) {
    array[i] = &values[n];
    n += n2;
  }

  double fac = 1.0/(n1*(n2+1));
  for (int i = 0; i < n1; i++) {
    for (int j = 0; j < n2; j++) {
      array[i][j] = fac*(i+1)*(j+1+1);
    }
  }

  return array;
}

void destroy_matrix(double** array) {
  delete [] array[0];
  delete [] array;
}
